daemon/guestos_mgr.c

```diff
@@ -371,30 +397,47 @@ push_config_verify_buf_cb(smartcard_crypto_verify_result_t verify_result, unsign
 	//			     GUESTOS_MGR_UPDATE_DOWNLOAD_NO_WIFI);
 
 	// 4. trigger image download if os is used by a container or a fresh install
-	if (guestos_mgr_is_guestos_used_by_containers(os_name) || !old_os) {
+	if (guestos_mgr_is_guestos_used_by_containers(os_name) || !old_os || cml_update) {
 		INFO("%s: %s", GUESTOS_MGR_UPDATE_TITLE, GUESTOS_MGR_UPDATE_DOWNLOAD);
-		guestos_mgr_download_latest(os_name);
+		if (control_send_message(CONTROL_RESPONSE_GUESTOS_MGR_INSTALL_STARTED, *resp_fd) < 0)
+			WARN("Could not send response to fd=%d", *resp_fd);
+		guestos_mgr_download_latest(os_name, *resp_fd);
 	}
+
+	mem_free(resp_fd);
 	return;
 
 cleanup_purge:
 	guestos_purge(os);
 cleanup_os:
 	guestos_free(os);
+err:
+	if (control_send_message(CONTROL_RESPONSE_GUESTOS_MGR_INSTALL_FAILED, *resp_fd) < 0)
+		WARN("Could not send response to fd=%d", *resp_fd);
+	mem_free(resp_fd);
 }
 
 int
 guestos_mgr_push_config(unsigned char *cfg, size_t cfglen, unsigned char *sig, size_t siglen,
-			unsigned char *cert, size_t certlen)
+			unsigned char *cert, size_t certlen, int resp_fd)
 {
 	INFO("Push GuestOS config (Phase 1)");
 
 	int res = -1;
+	int *cb_resp_fd = mem_new0(int, 1);
+	*cb_resp_fd = resp_fd;
+
 	if (cfg && sig && cert) {
 		res = smartcard_crypto_verify_buf(cfg, cfglen, sig, siglen, cert, certlen,
 						  GUESTOS_MGR_VERIFY_HASH_ALGO,
-						  push_config_verify_buf_cb, NULL);
+						  push_config_verify_buf_cb, cb_resp_fd);
 	}
+	if (res < 0) {
+		if (control_send_message(CONTROL_RESPONSE_GUESTOS_MGR_INSTALL_FAILED, resp_fd) < 0)
+			WARN("Could not send response to fd=%d", resp_fd);
+		mem_free(cb_resp_fd);
+	}
+
```

Comment:
free also if res >= 0?
(i think we should free arguments to callbacks outside the callee)

Reply:
nop, this is async. therefore we alloc memory for the data associated with the callback an leave it to the callback to free its resources when done. in the case res is < 0 the callback will not be triggered since an error happen during setup, thus we have to free here.